### ISSUE TITLE
[Discussion] Reach Router

### ISSUE THREAD

**Issue by bkniffler** [nozzle/react-static#750]
Hey there,

Reach router is a fairly new router by Ryan Florence that has a focus on [accessibility](https://reach.tech/router/accessibility), which is great for the type of apps that I create for react-static (mostly statically built frontends). It also comes with a smaller bundle size than react. Gatsby has switched already (https://github.com/gatsbyjs/gatsby/issues/5656). 

Would you be open for switching? I could PR the changes if I know that you like the idea. 

I also thought about react-static supporting multiple routers since, in the end, react-router is not used in lots of places. But I'm not sure if it really is worth the effort.

**Comment by tannerlinsley** [comment]
I would be open to this. I love reach router and I think it would make react-static a better library. 

Some thoughts:
- React-Static is very hands off with most things, but routing is currently very crucial to the guts of how react-static works. In a perfect world, we could support many routers, but I don't think that's a possibility right now. 
- This should be a change to the v6 branch, because it will be a breaking change
- We should abstract the router api to be more generic, or simply break the expectation that whatever router we use will be fully exposed via react-static. I think having our own set of simple components and methods would be much easier. This would also lay the groundwork to support multiple routers in the near future.
- Supporting multiple routers will be easier once Suspense and time-slicing are released.

**Comment by KidkArolis** [comment]
I was just thinking about this topic and was thinking about trying to integrate space-router into react-static, purely to bring the overall bundle size down. The routing doesn't need to be that heavy (maybe, maybe it does).

https://bundlephobia.com/result?p=react-router@4.3.1 8kb
https://bundlephobia.com/result?p=@reach/router@1.1.1 6kb
https://bundlephobia.com/result?p=space-router@0.7.0 1.8kb

Ideally, we should minimize the overhead of react-static to a minimum?

I would probably vote against supporting multiple routers though if that comes at a cost of how easy it is to use or maintain the lib.

**Comment by tannerlinsley** [comment]
Bundle size is very important, but so is usability, accessibility and SEO. IMO, reach-router is better equipped to fill the needs of the majority use case. If we're interested in cutting down on runtime bundle size, I think there are other areas we can focus on without sacrificing anything.

**Comment by KidkArolis** [comment]
Yeah, I get it! The accessibility stuff is really cool and I should look to add that to space-router anyway. SEO – not sure if router is relevant for that? Usability – wasn't gonna change the API that react-static exposes too much.

But yeah, I see your reasoning, react/reach router is definitely more battle tested.

**Comment by tannerlinsley** [comment]
I would really like to get this replaced and flesh out the routing API a lot more for v6:

My hopes for the api would be something like:
```javascript
import { Router, navigate } from 'react-static'

// Not much changes here
<Router />
// Support an optional history object override
<Router optionalHistory={history} />

// push
navigate('to/this/page')


```

**Comment by tannerlinsley** [comment]
I would really like to get this replaced and flesh out the routing API a lot more for v6:

My hopes for the api would be something like:
```javascript
import { Router, navigate } from 'react-static'

// Not much changes here
<Router />
// Support an optional history object override
<Router optionalHistory={history} />

// Push
navigate({ path: 'to/this/page' })
// Replace
navigate({ path: 'to/this/page', replace: true })
// With query params
navigate({ path: 'to/this/page', query: old => ({ ...old, foo: 'bar' }) })
// Just query params
navigate({ query: old => ({ ...old, foo: 'bar' }) })
// Replace just query params
navigate({ query: old => ({ ...old, foo: 'bar' }), replace: true })
// Support async/promise
await navigate(...)

```
